### PR TITLE
[Flax] Add DreamBooth

### PR DESCRIPTION
**Pull request by duongna21**
# What does this PR do?

Add Flax example for DreamBooth. 

How to run (74% faster than PyTorch example with same args on Tesla A100)
```
export MODEL_NAME="duongna/stable-diffusion-v1-4-flax"
export INSTANCE_DIR="path-to-instance-images"
export CLASS_DIR="path-to-class-images"
export OUTPUT_DIR="path-to-save-model"

python train_dreambooth_flax.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --instance_data_dir=$INSTANCE_DIR \
  --class_data_dir=$CLASS_DIR \
  --output_dir=$OUTPUT_DIR \
  --with_prior_preservation --prior_loss_weight=1.0 \
  --instance_prompt="a photo of sks dog" \
  --class_prompt="a photo of dog" \
  --resolution=512 \
  --train_batch_size=1 \
  --learning_rate=5e-6 \
  --num_class_images=200 \
  --max_train_steps=800
```
Prompt: `a photo of sks dog`

ảnh

## Who can review?
cc @patil-suraj @patrickvonplaten

### CODE REVIEW

**Comment by patil-suraj** [comment]
You are on fire @duongna21 !  🔥

File: examples/dreambooth/train_dreambooth_flax.py [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+        "--gradient_accumulation_steps",
+        type=int,
+        default=1,
+        help="Number of updates steps to accumulate before performing a backward/update pass.",
+    )
+    parser.add_argument(
+        "--gradient_checkpointing",
+        action="store_true",
+        help="Whether or not to use gradient checkpointing to save memory at the expense of slower backward pass.",
+    )
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:139]: These two features are not implemented for `flax`, let's remove these for now.

File: examples/dreambooth/train_dreambooth_flax.py [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+            'The scheduler type to use. Choose between ["linear", "cosine", "cosine_with_restarts", "polynomial",'
+            ' "constant", "constant_with_warmup"]'
+        ),
+    )
+    parser.add_argument(
+        "--lr_warmup_steps", type=int, default=500, help="Number of steps for the warmup in the lr scheduler."
+    )
+    parser.add_argument(
+        "--use_8bit_adam", action="store_true", help="Whether or not to use 8-bit Adam from bitsandbytes."
+    )
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:166]: This is not required in flax

File: examples/dreambooth/train_dreambooth_flax.py [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+        level=logging.INFO,
+    )
+    # Setup logging, we only want one process per machine to log things on the screen.
+    logger.setLevel(logging.INFO if jax.process_index() == 0 else logging.ERROR)
+    if jax.process_index() == 0:
+        transformers.utils.logging.set_verbosity_info()
+    else:
+        transformers.utils.logging.set_verbosity_error()
+
+    if args.seed is not None:
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:344]: think  `seed` should not be `None`

File: examples/dreambooth/train_dreambooth_flax.py [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+                rng = jax.random.split(rng)[0]
+                sample_rng = jax.random.split(rng, jax.device_count())
+                images = pipeline(prompt_ids, p_params, sample_rng, jit=True).images
+                images = images.reshape((images.shape[0] * images.shape[1],) + images.shape[-3:])
+                images = pipeline.numpy_to_pil(np.array(images))
+
+                for i, image in enumerate(images):
+                    hash_image = hashlib.sha1(image.tobytes()).hexdigest()
+                    image_filename = class_images_dir / f"{example['index'][i] + cur_class_images}-{hash_image}.jpg"
+                    image.save(image_filename)
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:398]: Very cool!

File: examples/dreambooth/train_dreambooth_flax.py [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+
+            with open(os.path.join(args.output_dir, ".gitignore"), "w+") as gitignore:
+                if "step_*" not in gitignore:
+                    gitignore.write("step_*\n")
+                if "epoch_*" not in gitignore:
+                    gitignore.write("epoch_*\n")
+        elif args.output_dir is not None:
+            os.makedirs(args.output_dir, exist_ok=True)
+
+    rng = jax.random.PRNGKey(args.seed)
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:363]: This will break if `seed` is `None`

File: examples/dreambooth/train_dreambooth_flax.py [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+
+    total_train_batch_size = args.train_batch_size * jax.local_device_count()
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=total_train_batch_size, shuffle=True, collate_fn=collate_fn
+    )
+
+    # Load models and create wrapper for stable diffusion
+    text_encoder = FlaxCLIPTextModel.from_pretrained("openai/clip-vit-large-patch14")
+    vae, vae_params = FlaxAutoencoderKL.from_pretrained(args.pretrained_model_name_or_path, subfolder="vae")
+    unet, unet_params = FlaxUNet2DConditionModel.from_pretrained(args.pretrained_model_name_or_path, subfolder="unet")
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:467]: Let's pass a `dtype` arg here, to enable `mixed_precision` training.

we can set `dtype` depending on the `--mixed_precision` argument.

**Review by patil-suraj — APPROVED** [review]
Looks very good, amazing work! Just left some comments.

Let's make sure that `seed` is not `None` as `PRNGKey` will break. Also let's update the readme to show hot to run this example. Then this should be good to merge :)

**Comment by patil-suraj** [comment]
For future PR, maybe also enable the option to allow training the `text_encoder` this has been found to improve results significantly.

**Comment by duongna21** [comment]
> (For future PR, maybe also enable the option to allow training the `text_encoder` this has been found to improve results significantly)

@patil-suraj Actually train_text_encoder has been allowed in this PR. Please check it out.
Also thank you for other very helpful comments. Addressed them!

File: examples/dreambooth/train_dreambooth_flax.py [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+    total_train_batch_size = args.train_batch_size * jax.local_device_count()
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset, batch_size=total_train_batch_size, shuffle=True, collate_fn=collate_fn, drop_last=True
+    )
+
+    weight_dtype = jnp.float32
+    if args.mixed_precision == "fp16":
+        weight_dtype = jnp.float16
+    elif args.mixed_precision == "bf16":
+        weight_dtype = jnp.bfloat16
```
Comment (patil-suraj) [inline examples/dreambooth/train_dreambooth_flax.py:451]: very cool!

File: examples/dreambooth/README.md [hunk 8]
```diff
@@ -58,6 +58,24 @@ accelerate launch train_dreambooth.py \
...
+export OUTPUT_DIR="path-to-save-model"
+
+python train_dreambooth_flax.py \
+  --pretrained_model_name_or_path=$MODEL_NAME  \
+  --instance_data_dir=$INSTANCE_DIR \
+  --output_dir=$OUTPUT_DIR \
+  --instance_prompt="a photo of sks dog" \
+  --resolution=512 \
+  --train_batch_size=1 \
+  --learning_rate=5e-6 \
```
Comment (patil-suraj) [inline examples/dreambooth/README.md:75]: Thanks for adding this!

**Review by patil-suraj — COMMENTED** [review]
Thanks a lot @duongna21 !

Also, it seems like `text_encoder` is always trained , no ? I think we should add an option called `--train_text_encoder` and train it only when its `True`. Because training `text_encoder` is not always needed.

**Comment by duongna21** [comment]
> Also, it seems like `text_encoder` is always trained? 

@patil-suraj No. You can ctrl + F for `if args.train_text_encoder` in the script and check if anything is wrong.

**Comment by patil-suraj** [comment]
Ahh, sorry. I missed that, all looks good now, thanks a lot for working on this!

**Review by patil-suraj — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Great work here!

File: examples/dreambooth/train_dreambooth_flax.py [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+
+    # Load models and create wrapper for stable diffusion
+    text_encoder = FlaxCLIPTextModel.from_pretrained("openai/clip-vit-large-patch14", dtype=weight_dtype)
```
Comment (skirsten) [inline examples/dreambooth/train_dreambooth_flax.py:454]: Hi, why is this version pulled from the hub and its not just using the one in the `text_encoder` subfolder in `args.pretrained_model_name_or_path`?

Is there something wrong with the other version?

Comment (skirsten) [reply examples/dreambooth/train_dreambooth_flax.py:454]: @douwekiela Awesome! Thanks for explaining it

Comment (pcuenca) [reply examples/dreambooth/train_dreambooth_flax.py:454]: @patil-suraj yes, it's merged

Comment (patil-suraj) [reply examples/dreambooth/train_dreambooth_flax.py:454]: Awesome, @duongna21 would you like to open a PR and to update this then :)  We will also need to update the installation instructions for `transformers` to include that fix
